aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/EndpointBuiltInsDecorator.kt

```diff
@@ -0,0 +1,179 @@
+            return listOfNotNull(
+                codegenContext.model.sdkConfigSetter(codegenContext.serviceShape.id, builtIn, clientParam?.name),
+            )
+        }
+
+        override fun configCustomizations(
+            codegenContext: ClientCodegenContext,
+            baseCustomizations: List<ConfigCustomization>,
+        ): List<ConfigCustomization> {
+            return baseCustomizations.extendIf(rulesetContainsBuiltIn(codegenContext)) {
+                standardConfigParam(
+                    clientParam ?: toConfigParam(builtIn),
+                )
+            }
+        }
+
+        override fun endpointCustomizations(codegenContext: ClientCodegenContext): List<EndpointCustomization> = listOf(
+            object : EndpointCustomization {
+                override fun loadBuiltInFromServiceConfig(parameter: Parameter, configRef: String): Writable? =
+                    when (parameter.builtIn) {
+                        builtIn.builtIn -> writable {
+                            rust("$configRef.$name")
+                            if (parameter.type == ParameterType.STRING) {
+                                rust(".clone()")
+                            }
+                        }
+                        else -> null
+                    }
+
+                override fun setBuiltInOnServiceConfig(name: String, value: Node, configBuilderRef: String): Writable? {
+                    if (name != builtIn.builtIn.get()) {
+                        return null
+                    }
+                    return writable {
+                        rustTemplate(
+                            "let $configBuilderRef = $configBuilderRef.${nameOverride ?: builtIn.name.rustName()}(#{value});",
+                            "value" to value.toWritable(),
+                        )
+                    }
+                }
+            },
+        )
+    }
+}
+
+private val endpointUrlDocs = writable {
+    rust(
+        """
+        /// Sets the endpoint url used to communicate with this service
+
+        /// Note: this is used in combination with other endpoint rules, e.g. an API that applies a host-label prefix
+        /// will be prefixed onto this URL. To fully override the endpoint resolver, use
+        /// [`Builder::endpoint_resolver`].
+        """.trimIndent(),
+    )
+}
+
+fun Node.toWritable(): Writable {
+    val node = this
+    return writable {
+        when (node) {
+            is StringNode -> rust(node.value.dq())
+            is BooleanNode -> rust("${node.value}")
+            else -> PANIC("unsupported value for a default: $node")
+        }
+    }
+}
+
+val PromotedBuiltInsDecorators =
+    listOf(
+        decoratorForBuiltIn(Builtins.FIPS),
+        decoratorForBuiltIn(Builtins.DUALSTACK),
+        decoratorForBuiltIn(
+            Builtins.SDK_ENDPOINT,
+            ConfigParam("endpoint_url", RuntimeType.String.toSymbol(), endpointUrlDocs),
+        ),
```

Comment:
If a customer is using a custom smithy client with SDK codegen disabled, can they easily set endpoints? This looks SDK specific, and I'm wondering how it interacts with the "plain" smithy client codegen.